Source/Details/ASPageTable.mm

```diff
@@ -0,0 +1,146 @@
+{
+  CGRect contentRect = CGRectMake(0.0, 0.0, contentSize.width, contentSize.height);
+  // Make sure the specified rect is within contentRect
+  rect = CGRectIntersection(rect, contentRect);
+  if (CGRectIsNull(rect) || CGRectIsEmpty(rect)) {
+    return nil;
+  }
+  
+  NSPointerArray *results = [NSPointerArray pointerArrayWithOptions:(NSPointerFunctionsIntegerPersonality | NSPointerFunctionsOpaqueMemory)];
+  
+  ASPageCoordinate minPage = ASPageCoordinateForPageThatContainsPoint(CGPointMake(CGRectGetMinX(rect), CGRectGetMinY(rect)), pageSize);
+  ASPageCoordinate maxPage = ASPageCoordinateForPageThatContainsPoint(CGPointMake(CGRectGetMaxX(rect), CGRectGetMaxY(rect)), pageSize);
+  if (minPage == maxPage) {
+    [results addPointer:(void *)minPage];
+    return results;
+  }
+  
+  NSUInteger minX = ASPageCoordinateGetX(minPage);
+  NSUInteger minY = ASPageCoordinateGetY(minPage);
+  NSUInteger maxX = ASPageCoordinateGetX(maxPage);
+  NSUInteger maxY = ASPageCoordinateGetY(maxPage);
+  
+  for (NSUInteger x = minX; x <= maxX; x++) {
+    for (NSUInteger y = minY; y <= maxY; y++) {
+      ASPageCoordinate page = ASPageCoordinateMake(x, y);
+      [results addPointer:(void *)page];
+    }
+  }
+  
+  return results;
+}
+
+@implementation NSMapTable (ASPageTableMethods)
+
++ (instancetype)pageTableWithValuePointerFunctions:(NSPointerFunctions *)valueFuncs
+{
+  static NSPointerFunctions *pageCoordinatesFuncs;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    pageCoordinatesFuncs = [NSPointerFunctions pointerFunctionsWithOptions:NSPointerFunctionsIntegerPersonality | NSPointerFunctionsOpaqueMemory];
+  });
+  
+  return [[NSMapTable alloc] initWithKeyPointerFunctions:pageCoordinatesFuncs valuePointerFunctions:valueFuncs capacity:0];
+}
+
++ (ASPageTable *)pageTableForStrongObjectPointers
+{
+  static NSPointerFunctions *strongObjectPointerFuncs;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    strongObjectPointerFuncs = [NSPointerFunctions pointerFunctionsWithOptions:NSPointerFunctionsStrongMemory];
+  });
+  return [self pageTableWithValuePointerFunctions:strongObjectPointerFuncs];
+}
+
++ (ASPageTable *)pageTableForWeakObjectPointers
+{
+  static NSPointerFunctions *weakObjectPointerFuncs;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    weakObjectPointerFuncs = [NSPointerFunctions pointerFunctionsWithOptions:NSPointerFunctionsWeakMemory];
+  });
+  return [self pageTableWithValuePointerFunctions:weakObjectPointerFuncs];
+}
+
++ (ASPageTable *)pageTableWithLayoutAttributes:(id<NSFastEnumeration>)layoutAttributesEnumerator pageSize:(CGSize)pageSize
+{
+  ASPageTable *result = [ASPageTable pageTableForStrongObjectPointers];
+  for (UICollectionViewLayoutAttributes *attrs in layoutAttributesEnumerator) {
+    ASPageCoordinate page = ASPageCoordinateForPageThatContainsPoint(attrs.frame.origin, pageSize);
```

Comment:
I think we need to key this on page-that-intersects-rect. You can imagine an item that spans 10 pages – we need to make sure we include that one whenever we search any of those 10 pages.